### ISSUE TITLE
[ONNX] when exporting a PyTorch model, creates an unnecessary variable that messes up inferencing.

### ISSUE THREAD

**Issue by ugurkc** [pytorch/pytorch#84514]
### 🐛 Describe the bug

I have a NLP application I've built on PyTorch w/ transformers RoBERTa. As expected, the inferencing should take 2 parameters, input_ids & attention_mask.  However, if I export the said model and view it on netron, it shows 3 inputs: input_ids,attention_mask, and a onnx::Reshape_2 input that shouldn't be there. 

My personal prediction is that because my forward pass is implemented as follows:
```python
    def forward(self, input_ids, attention_mask, labels=None):
      # roberta layer
      output = self.pretrained_model(input_ids=input_ids, attention_mask=attention_mask)
      pooled_output = torch.mean(output.last_hidden_state, 1)
      # final logits
      pooled_output = self.dropout(pooled_output)
      pooled_output = self.hidden(pooled_output)
      pooled_output = F.relu(pooled_output)
      pooled_output = self.dropout(pooled_output)
      logits = self.classifier(pooled_output)
      # calculate loss
      loss = 0
      if labels is not None:
        loss = self.loss_func(logits.view(-1, self.config['n_labels']), labels.view(-1, self.config['n_labels']))
      return loss, logits
```
Onnx converter thinks that labels, is another input that's required for inferencing. But it's not named as "labels" aswell, so that is also confusing. 
After renaming this onnx:Reshape_2 as "labels" , If, one tries to take this model, and run it with an inferencing script in another python environment, we get the following error:

The script:
```python
import onnx
import torch
import numpy as np
from transformers import AutoTokenizer



onnx_model = onnx.load('ekman-v1-torch (1).onnx')

text = "Text from the news article"

tokenizer = AutoTokenizer.from_pretrained("roberta-base")
tokens = tokenizer.encode_plus(text,
                    add_special_tokens = True,
                    return_tensors = 'np',
                    truncation = True,
                    max_length = 128,
                    padding = 'max_length',
                    return_attention_mask = True
                    )

input_ids = tokens.input_ids
attention_mask = tokens.attention_mask
labels = np.zeros(6).astype(np.float32)

print(labels)

input ={'input_ids' : input_ids,
        'attention_mask' : attention_mask,
        'labels' : labels} 



onnx.checker.check_model(onnx_model)

import onnxruntime as ort
import numpy as np


ort_sess = ort.InferenceSession('ekman-v1-torch (1).onnx')
outputs = ort_sess.run(None, {'input_ids' : input_ids,'attention_mask' : attention_mask,'labels': labels})
```

Error Message:

```
  File "/Users/MyUser/Library/Python/3.8/lib/python/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 200, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Invalid rank for input: labels Got: 1 Expected: 2 Please fix either the inputs or the model.
```

This error basicly says, that my input of 6 zeros (a dummy version of my 6 classed labels) is invalid, and needs 2 , if I substitute: labels = np.zeros(6).astype(np.float32) with labels = np.zeros((2,6)).astype(np.float32) then the error says

```
  File "/Users/ugurkoc/Library/Python/3.8/lib/python/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 200, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: labels for the following indices
 index: 0 Got: 2 Expected: 1
```

When I inspect the problem with netron, it shows an input right at the end of the computational graph, that seems useless as predicted to the whole graph. 

<img width="1506" alt="Screen Shot 2022-09-03 at 13 38 13" src="https://user-images.githubusercontent.com/32737088/188269883-2dc32f1d-1c16-45ee-931b-ccf7dca76f60.png">

Finally if I were to remove this "labels" third input from my model seeing it's useless, then the error message pops up saying:

[ CPULongType{} ]) of traced region did not have observable data dependence with trace inputs; this probably indicates your program cannot be understood by the tracer.

Thanks in advance.

### Versions

Collecting environment information...
PyTorch version: 1.12.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.6
Libc version: glibc-2.26

Python version: 3.7.13 (default, Apr 24 2022, 01:04:09)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.188+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] pytorch-lightning==1.7.4
[pip3] torch==1.12.1+cu113
[pip3] torchaudio==0.12.1+cu113
[pip3] torchmetrics==0.9.3
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.13.1
[pip3] torchvision==0.13.1+cu113

**Comment by ugurkc** [comment]
Here is how I create my export aswell:

```
X = next(iter(dummy_dl))
del X["labels"]

print(X)

torch.onnx.export(loaded_model,X,"ekman-v1-torch.onnx",export_params = True, do_constant_folding = True, input_names = ['input_ids','attention_mask','labels'],output_names = [])
```

**Comment by AllenTiTaiWang** [comment]
> Here is how I create my export aswell:
> 
> ```python
> X = next(iter(dummy_dl))
> del X["labels"]
> 
> print(X)
> 
> torch.onnx.export(loaded_model,X,"ekman-v1-torch.onnx",export_params = True, do_constant_folding = True, input_names = ['input_ids','attention_mask','labels'],output_names = [])
> ```

Hi @ugurkc ,

`torch.onnx.export` provides **dynamic_axes** for user to keep their calculation with dynamic shapes. I think that might be one of the reasons you are getting a weird graph. Please find the usage of it in this link: https://pytorch.org/docs/stable/onnx.html#torch.onnx.export

**Comment by justinchuby** [comment]
I would suggest extracting the loss computation part out from your forward function (so that it is not part of your model), as it is not needed for inference and should not be traced during export.

**Comment by ugurkc** [comment]
Thank you very much for comments

- @AllenTiTaiWang: the dynamic_axes implementation solved my issue. 👍 
- @justinchuby: Noted, it was in my TODO list for a while now. 👍
